scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3RawChange.java

```diff
@@ -55,22 +56,7 @@ public ByteBuffer getUnsafeBytes(ColumnDefinition c) {
 
     @Override
     public String toString() {
-        StringBuilder result = new StringBuilder();
-        result.append("Driver3RawChange(");
-        boolean first = true;
-        for (ChangeSchema.ColumnDefinition columnDefinition : getSchema().getAllColumnDefinitions()) {
-            if (!first) {
-                result.append(",");
-            } else {
-                first = false;
-            }
-
-            String name = columnDefinition.getColumnName();
-            Cell cell = getCell(name);
-            result.append(name).append("=").append(cell);
-        }
-        result.append(")");
-        return result.toString();
+        return stream().map(Object::toString).collect(Collectors.joining(", ", "RawChange(", ")"));
```

Comment:
Why did you change Driver3RawChange( to RawChange(? It seems more correct to state the exact type, no?

Reply:
No. The user does not know which implementation it is, and a driver4 impl should give the same output imho, If it as not for how inheritance works, it could be a defaulted method in the interface.

Reply:
I don't think driver3 and driver4 have to return the same String. That's the purpose of overriding the method. toString is mostly used for logging/debugging so it's better to be more specific than not. toString is not an element of RawChange interface. It's an element of Object interface and its javadoc says:
```
Returns a string representation of the object.
In general, the {@code toString} method returns a string that "textually represents" this object.
```
It explicitly says **this** object not the superclass or anything else.

Reply:
No, it says "this" object as in not another object. A toString should describe an object in a manner useful to the user. While not a pretty-print, it should be usable as a printf debugging tool. That means for data carrier interfaces like this, representing its contents, not its implementation type.

Reply:
I'm not fully convinced. Let's take an example from Swing:
superclass AbstractElement -> https://github.com/himnay/java7-sourcecode/blob/329bbb33cbe8620aee3cee533eec346b4b56facd/javax/swing/text/AbstractDocument.java#L1764
one subclass BranchElement -> https://github.com/himnay/java7-sourcecode/blob/329bbb33cbe8620aee3cee533eec346b4b56facd/javax/swing/text/AbstractDocument.java#L2302
another subclass LeftElement -> https://github.com/himnay/java7-sourcecode/blob/329bbb33cbe8620aee3cee533eec346b4b56facd/javax/swing/text/AbstractDocument.java#L2519
they seem to be very specific in toString

Reply:
That is because their types carries important info - namely what type of node it is (leaf, branch etc...)

Reply:
So do Driver3RawChange and potential Driver4RawChange. Adding class name to toString seems to be a common practice. For example java.beans.PropertyChangeEvent will include the name of every subclass in its toString -> https://github.com/himnay/java7-sourcecode/blob/329bbb33cbe8620aee3cee533eec346b4b56facd/java/beans/PropertyChangeEvent.java#L151

Reply:
Because it is the only aspect of naming in the hierarchy that carries meaning. In the case of Driver3RawChange, the one that carries meaning is instead RawChange. But I am to tired to debate. Now all printouts will carry pointless filler info and be harder to read.

Reply:
Thank you for making the codebase "worse" :)

Reply:
It is what I do. Wait, that did not sound right...